### PR TITLE
Make pure menu dropdowns keyboard accessible

### LINKED ISSUE #67: "Rust ▾" menu is not keyboard accessible
Tab into it, then press enter. Nothing happens.

### PR DESCRIPTION
**Pull request by notriddle**
Fixes #67

### CODE REVIEW

**Comment by jyn514** [comment]
This is an awful lot of JavaScript... Is there a way to do this with CSS instead?

**Comment by pietroalbini** [comment]
I'm fine adding JavaScript to improve accessibility (if it can't be done with HTML itself), but it should be in a separate JavaScript style: we'll soon want to add a CSP, and if we do that we won't enable inline JS.

**Comment by GuillaumeGomez** [comment]
Also, it's not working really well...

**Comment by notriddle** [comment]
> This is an awful lot of JavaScript... Is there a way to do this with CSS instead?

Parts of this — binding to the escape key and arrow keys, and focus trapping — are impossible with just CSS. There are some nasty hacks that you can do with checkboxes to at least make opening the menus possible, but it would require us to modify the markup and basically fork Pure.css to do it.

So, to get something truly *good*, you need some JavaScript anyway.

> Also, it's not working really well...

... like WHAT?!

**Comment by jyn514** [comment]
> Parts of this — binding to the escape key and arrow keys, and focus trapping — are impossible with just CSS. There are some nasty hacks that you can do with checkboxes to at least make opening the menus possible, but it would require us to modify the markup and basically fork Pure.css to do it.

I see. I took a look at the [purecss docs](https://purecss.io/menus/#dropdowns) and it looks like they have a recommended javascript solution: https://purecss.io/js/menus.js. Could we use that instead instead of writing something from scratch?

**Comment by notriddle** [comment]
> Could we use that instead instead of writing something from scratch?

Doesn't work right with the crate mega-dropdown. That's why I wrote a custom one.

**Comment by notriddle** [comment]
@pietroalbini I pushed a commit to move the menu script to a separate file. It's also minified.

**Comment by GuillaumeGomez** [comment]
It's still not working for me?

**Comment by notriddle** [comment]
@GuillaumeGomez Okay, after running through a few more tests, I think I've found the nature of your problem. Probably.

I hadn't been using the Docker scripts (actually, I've just been using the Firefox Developer Tools to inject my script into the actual docs.rs site). Now that I tried using them, I found that I had failed to update the Dockerfile. I've fixed that.

**Comment by GuillaumeGomez** [comment]
No, I meant the JS code. I ran it into my browser directly.

**Comment by notriddle** [comment]
> No, I meant the JS code. I ran it into my browser directly.

So when you do that, is anything written to the console? What browser are you using? Are you just mousing over, clicking, or pressing Enter on the menu button? I would like you to write detailed instructions for reproducing your problem.

**Comment by GuillaumeGomez** [comment]
No, the code is fine apparently. But when clicking on the dropdown menu, I randomly get a grey background all over the page and I have to click again to me it disappear. That's definitely not very nice. :-/

**Comment by notriddle** [comment]
> No, the code is fine apparently. But when clicking on the dropdown menu, I randomly get a grey background all over the page and I have to click again to me it disappear.

If other people also dislike it, I'm fine with removing the backdrop thing. Does anybody else have an opinion on it?

**Comment by jyn514** [comment]
I'm not a fan of focus-stealing in general.

**Comment by notriddle** [comment]
Okay, the gray background is now gone.

File: templates/menu.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
+// Allow menus to be open and used by keyboard.
+(function() {
+    var currentMenu;
+    var backdrop = document.createElement("div");
```
[the remark this answers is not in the document]
Comment (notriddle) [reply templates/menu.js]: Yes. It's not visible, but it's used to implement the behaviour where clicking outside the menu dismisses it.

File: templates/menu.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        currentMenu = null;
+        backdrop.style.display = "none";
+    }
+    backdrop.onclick = closeMenu;
+    function openMenu(newMenu) {
+        currentMenu = newMenu;
+        newMenu.className += " pure-menu-active";
+        backdrop.style.display = "block";
+    }
+    function menuOnClick(e) {
```
[the remark this answers is not in the document]
Comment (notriddle) [reply templates/menu.js]: I don't know of any way to have it respond on space, but I can just bind the keyboard.

**Comment by GuillaumeGomez** [comment]
Did you try with `tabindex` tag by any chance?

File: templates/menu.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                    // so they won't do it on accident.
+                    switchTo = allItems[0];
+                    break;
+                case "end":
+                    // end: focus last menu item.
+                    // This is the behavior of WAI, while GitHub scrolls,
+                    // but it's unlikely that a user will try to scroll the page while the menu is open,
+                    // so they won't do it on accident.
+                    switchTo = last(allItems);
+                    break;
```
[the remark this answers is not in the document]
Comment (notriddle) [reply templates/menu.js]: What should pressing page up and page down do? On my system (linux mint) pressing them while I've got the menu open does nothing. Should it just swallow them? Or should they seek to the top and bottom of the menu?

**Comment by notriddle** [comment]
> Did you try with `tabindex` tag by any chance?

What part of this functionality would I use `tabindex` to implement?

**Comment by pietroalbini** [comment]
By the way, just tested this locally, and it works great! Thanks for working on this!

File: templates/menu.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply templates/menu.js]: Seeking to the top and bottom is fine. Normally it scrolls by a set amount (~10-15 items) but I don't think we have enough menu options for it to make a difference.

**Comment by GuillaumeGomez** [comment]
@notriddle The point was to make it work with the keyboard, no? With `tabindex`, it's selectable using TAB.

**Comment by notriddle** [comment]
@GuillaumeGomez 

> The point was to make it work with the keyboard, no? With tabindex, it's selectable using TAB.

It already is selectable using TAB. The problem is making the menu *open* when you press DOWN on it.

**Comment by jyn514** [comment]
Looks great to me! @GuillaumeGomez What's this about tabindex?

File: Cargo.lock [hunk 4]
```diff
@@ -327,6 +327,7 @@ dependencies = [
  "libc 0.2.62 (registry+https://github.com/rust-lang/crates.io-index)",
  "log 0.4.6 (registry+https://github.com/rust-lang/crates.io-index)",
  "magic 0.12.2 (registry+https://github.com/rust-lang/crates.io-index)",
+ "minifier 0.0.1 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (jyn514) [inline Cargo.lock]: One last nit: can you remove the changes to `Cargo.lock`?

File: build.rs [hunk 5]
```diff
@@ -49,3 +50,18 @@ fn compile_sass() {
...
+    let source_path = Path::new(concat!(env!("CARGO_MANIFEST_DIR"), "/templates/menu.js"));
+    let dest_path = Path::new(&env::var("OUT_DIR").unwrap()).join("menu.js");
+    File::open(&source_path)
+        .expect("open templates/menu.js")
+        .read_to_string(&mut javascript)
+        .expect("read templates/menu.js");
+    File::create(&dest_path)
+        .expect("create target/.../menu.js")
+        .write_all(javascript.as_bytes())
+        .expect("write target/.../menu.js");
```
Comment (jyn514) [inline build.rs]: nit: This might be easier using [`fs::copy`](https://doc.rust-lang.org/std/fs/fn.copy.html).

File: build.rs [hunk 6]
```diff
@@ -49,3 +50,15 @@ fn compile_sass() {
...
+
+fn copy_js() {
+    use std::io::Read;
+    let mut javascript = String::new();
+    let source_path = Path::new(concat!(env!("CARGO_MANIFEST_DIR"), "/templates/menu.js"));
+    let dest_path = Path::new(&env::var("OUT_DIR").unwrap()).join("menu.js");
+    File::open(&source_path)
+        .expect("open templates/menu.js")
+        .read_to_string(&mut javascript)
+        .expect("read templates/menu.js");
```
Comment (jyn514) [inline build.rs]: You can now remove this call to `File::open` and the `javascript` variable.

(sorry I'm being so picky, the PR is great overall)

**Comment by jyn514** [comment]
Just deployed, this looks great :)

**Pull request merged**